flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxImpl.java

```diff
@@ -158,44 +174,108 @@ public boolean tryPutMail(@Nonnull Runnable letter) {
 	}
 
 	@Override
-	public void putMail(@Nonnull Runnable letter) throws InterruptedException {
+	public void putMail(@Nonnull Runnable letter) throws InterruptedException, MailboxStateException {
```

Comment:
Shouldn't we add a checkstate that we are not putting something from the mailbox thread?

Reply:
From how the mailbox is threated in the PR, it is more of a datastructure to track runnable objects and acts somewhat like a queue. Not sure if I would implement knowledge about threads and what threads are allowed to do what here. This is now implemented on the level of the executor, i.e. the user-facing facade.